ansible_collections/arista/avd/plugins/action/validate_and_template.py

```diff
@@ -0,0 +1,168 @@
+        if not isinstance(self.validation_mode, str):
+            raise AnsibleActionFail("The argument 'validation_mode' must be a string")
+        if self.validation_mode not in VALID_VALIDATION_MODES:
+            raise AnsibleActionFail(f"Invalid value '{self.validation_mode}' for the argument 'validation_mode'. Must be one of {VALID_VALIDATION_MODES}")
+
+        self.add_md_toc = self._task.args.get("add_md_toc", False)
+        if not isinstance(self.add_md_toc, bool):
+            raise AnsibleActionFail("The argument 'add_md_toc' must be a boolean")
+
+        self.md_toc_skip_lines = self._task.args.get("md_toc_skip_lines", 0)
+        if not isinstance(self.md_toc_skip_lines, int):
+            raise AnsibleActionFail("The argument 'md_toc_skip_lines' must be an integer")
+
+        # Build data from hostvars and role default vars
+        self.hostname = task_vars["inventory_hostname"]
+        self.data = self._templar.template(self._task._role.get_default_vars())
+        self.data.update(task_vars["hostvars"].get(self.hostname))
+
+        try:
+            self.avd_schema = AvdSchema(schema)
+        except Exception as e:
+            raise AnsibleActionFail("Invalid Schema supplied to the 'arista.avd.validate_and_template' plugin") from e
+
+        result_messages = []
+
+        # Perform data conversions
+        if self.conversion_mode != "disabled":
+            self.convert_data(result_messages)
+
+        # Perform validation
+        if self.validation_mode != "disabled":
+            valid = self.validate_data(result_messages)
+            if not valid and self.validation_mode == "error":
+                result["failed"] = True
+
+        if result_messages:
+            result["msg"] = " ".join(result_messages)
+
+        # Template to file
+        # Update result from Ansible "copy" operation (setting 'changed' flag accordingly)
+        result.update(self.template(task_vars))
+
+        return result
+
+    def convert_data(self, result_messages):
+        """
+        Convert data according to the schema (convert_types)
+
+        The data conversion is done in-place (updating the original "data" dict).
+        """
+
+        # avd_schema.convert returns a generator, which we need to run over to perform the actual conversions.
+        # The returned values are a list of conversion errors
+        exceptions = list(self.avd_schema.convert(self.data))
+        conversion_counter = self.handle_exceptions(exceptions, self.conversion_mode)
+        # This is not yet effective, since the converter does not yield on succesful conversion.
+        # TODO: Get converter to return the data that was converted, so the user can see what needs to be updated.
+        if conversion_counter:
+            result_messages.append(f"{conversion_counter} data conversions done to conform to schema.")
+        return
+
+    def validate_data(self, result_messages):
```

Comment:
similarly for this one we could return result_messages or empty and test on this info if this is valid or not

Reply:
That would just add more logic to the calling code instead of just updating the original value.